src/mscorlib/src/System/Array.cs

```diff
@@ -1566,24 +1566,24 @@ public static void Reverse<T>(T[] array, int index, int length)
         {
             if (array == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
-            if (index < 0)
-                ThrowHelper.ThrowIndexArgumentOutOfRange_NeedNonNegNumException();
-            if (length < 0)
-                ThrowHelper.ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum();
-            if (array.Length - index < length)
+            if ((uint)index > (uint)array.Length)
+                ThrowHelper.ThrowArgumentOutOfRange_IndexException();
+            if ((uint)length > (uint)(array.Length - index))
                 ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
 
-            ref T p = ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData());
-            int i = index;
-            int j = index + length - 1;
-            while (i < j)
+            if (length <= 1)
+                return;
+
+            ref T first = ref Unsafe.Add(ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData()), index);
+            ref T last = ref Unsafe.Add(ref Unsafe.Add(ref first, length), -1);
```

Comment:
The reason for the two `Unsafe.Add` calls is that if we write `Unsafe.Add(ref first, length - 1)`, RyuJIT generates `dec length; movsxd length64, length; lea last, [first + length64 * stride];`. By using a second `Unsafe.Add`, RyuJIT can detect that the -1 is a constant, so it avoids emitting the first `dec` instruction and instead emits `lea last, [first + length64 * stride - stride]` as its last instruction.